<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Analyse the output of polyApipe.py — do_pipeline • polyApiper</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Analyse the output of polyApipe.py — do_pipeline" />
<meta property="og:description" content="Analyse the output of polyApipe.py. The resulting directory can be loaded
with load_banquet. An HTML report is also produced." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">polyApiper</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Analyse the output of polyApipe.py</h1>
    
    <div class="hidden name"><code>do_pipeline.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Analyse the output of polyApipe.py. The resulting directory can be loaded
with <code>load_banquet</code>. An HTML report is also produced.</p>
    </div>

    <pre class="usage"><span class='fu'>do_pipeline</span><span class='op'>(</span>
  <span class='va'>out_path</span>,
  counts_file_dir <span class='op'>=</span> <span class='cn'>NULL</span>,
  counts_files <span class='op'>=</span> <span class='cn'>NULL</span>,
  batch_names <span class='op'>=</span> <span class='st'>""</span>,
  <span class='va'>peak_info_file</span>,
  <span class='va'>organism</span>,
  cell_name_func <span class='op'>=</span> <span class='kw'>function</span><span class='op'>(</span><span class='va'>batch</span>, <span class='va'>cell</span><span class='op'>)</span> <span class='fu'><a href='https://rdrr.io/r/base/paste.html'>paste0</a></span><span class='op'>(</span><span class='va'>batch</span>, <span class='va'>cell</span><span class='op'>)</span>,
  cells_to_use <span class='op'>=</span> <span class='cn'>NULL</span>,
  remove_mispriming <span class='op'>=</span> <span class='cn'>TRUE</span>,
  utr_or_extension_only <span class='op'>=</span> <span class='cn'>FALSE</span>,
  peak_min_present <span class='op'>=</span> <span class='fl'>50</span>,
  peak_min_prop <span class='op'>=</span> <span class='fl'>0.01</span>,
  do_logNormCounts <span class='op'>=</span> <span class='cn'>TRUE</span>,
  do_computeSumFactors <span class='op'>=</span> <span class='cn'>TRUE</span>,
  title <span class='op'>=</span> <span class='st'>"polyApiper pipeline run"</span>,
  stages <span class='op'>=</span> <span class='fl'>1</span><span class='op'>:</span><span class='fu'><a href='https://rdrr.io/r/base/length.html'>length</a></span><span class='op'>(</span><span class='va'>PIPELINE_STAGES</span><span class='op'>)</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>out_path</th>
      <td><p>Output directory name.</p></td>
    </tr>
    <tr>
      <th>counts_file_dir</th>
      <td><p>A directory containing counts files. Batch names are taken from the basename of the count files without the .tab.gz suffix. Give either this argument or <code>counts_files</code> but not both.</p></td>
    </tr>
    <tr>
      <th>counts_files</th>
      <td><p>Alternative to counts_file_dir. One or more filenames for .tab.gz files produced by polyApipe.py. Give either this argument or <code>counts_file_dir</code> but not both.</p></td>
    </tr>
    <tr>
      <th>batch_names</th>
      <td><p>If using counts_files, give a vector of the batch/sample names, in same order as counts_files</p></td>
    </tr>
    <tr>
      <th>peak_info_file</th>
      <td><p>GTF formatted peak file as output from polyApipe.py.</p></td>
    </tr>
    <tr>
      <th>organism</th>
      <td><p>Organism directory, as created by <code>do_ensembl_organism</code>.</p></td>
    </tr>
    <tr>
      <th>cells_to_use</th>
      <td><p>Character vector of cells to use.</p></td>
    </tr>
    <tr>
      <th>remove_mispriming</th>
      <td><p>Remove peaks considered to be mispriming peaks.</p></td>
    </tr>
    <tr>
      <th>utr_or_extension_only</th>
      <td><p>Remove peaks in the 5'UTR, exons, or introns.</p></td>
    </tr>
    <tr>
      <th>peak_min_present</th>
      <td><p>A peak is retained if it is present in this number of cells.</p></td>
    </tr>
    <tr>
      <th>peak_min_prop</th>
      <td><p>A peak is used for APA and gene expression calculations only if it constitutes at least this proportion of total UMIs for the gene.</p></td>
    </tr>
    <tr>
      <th>do_logNormCounts</th>
      <td><p>Compute logcounts for gene and peak expression?</p></td>
    </tr>
    <tr>
      <th>do_computeSumFactors</th>
      <td><p>When computing logcounts, use scran::computeSumFactors? If not, unadjusted cell library sizes are used.</p></td>
    </tr>
    <tr>
      <th>title</th>
      <td><p>Title to use in report.</p></td>
    </tr>
    <tr>
      <th>stages</th>
      <td><p>Stages of the pipeline to run. Stages are listed in PIPELINE_STAGES.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>There is no return value. Results are placed in the <code>out_path</code> directory. They can be loaded using <code><a href='load_banquet.html'>load_banquet(out_path)</a></code>.</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>You can choose to run only specific stages of the pipeline with the <code>stages</code> argument. See the global <code>PIPELINE_STAGES</code> for possible stages. <code>stages</code> is a character vector or an integer vector indexing into <code>PIPELINE_STAGES</code>.</p>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Paul Harrison, Sarah Williams.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


