---
params:
    title: "polyApipe results overview"
    input: "NULL"
    wd: "."
output: 
    html_document:
        code_folding: hide
title: "`r params$title`"
---

```{r echo=F}
knitr::opts_knit$set(root.dir=params$wd)
knitr::opts_chunk$set(dpi=100, fig.width=6, fig.height=4)
```

```{r message=F}
library(SummarizedExperiment)
library(dplyr)
library(ggplot2)
library(tibble)

library(polyApiper)
library(weitrix)


banq <- load_banquet(params$input)

n_cells <- ncol(banq$raw_peak_counts)

raw_row_data <- as.data.frame(rowData(banq$raw_peak_counts))
raw_col_data <- as.data.frame(colData(banq$raw_peak_counts))
raw_col_data$total_umis <- colSums(assay(banq$raw_peak_counts,"counts"))

n_mispriming_peaks <- sum(raw_row_data$misprime)

n_input_peaks <- nrow(banq$raw_peak_counts)
n_used_peaks <- nrow(banq$peak_expr)
n_used_genes <- nrow(banq$gene_expr)
n_apa_peaks <- nrow(banq$prop)
n_apa_genes <- nrow(banq$shift)
```

`r strftime(Sys.time(),"%Y-%m-%d")`

<div style="font-size: 150%">
`r n_cells` cells used.

`r n_input_peaks` peaks detected. `r n_mispriming_peaks` peaks regarded as mispriming.

Peak breakdown:
</div>

```{r}
raw_row_data %>% 
    mutate(region=ifelse(is.na(region),"unknown",region)) %>%
    mutate(region=factor(region, c("intron","exon","3'UTR","extension","unknown"))) %>%
    dplyr::count(misprime, region) %>% 
    spread(region, n, fill=0) %>%
    knitr::kable()
```

## Used in analysis

<div style="font-size: 150%">
After filtering steps, `r n_used_peaks` peaks in `r n_used_genes` genes.

APA of `r n_apa_genes` genes measured using `r n_apa_peaks` peaks.
</div>

```{r}
raw_col_data %>%
    rownames_to_column("rowname") %>% 
    mutate(used = rowname %in% colnames(banq$gene_expr$weitrix)) %>%
    arrange(-total_umis) %>%
    rowid_to_column("rowid") %>%
    arrange(used) %>%
    ggplot(aes(x=rowid, y=total_umis)) + 
    geom_point() +
    scale_y_log10() +
    labs(x="Cell (largest to smallest)", y="Total reads in peaks")
```

## Versions

```{r}
sessionInfo()
```


